File: performance-buffered-reader/src/test/java/com/jesperancinha/test/objects/bufferedreader/NanoStabilityCheckerTest.java
package com.jesperancinha.test.objects.bufferedreader;

import static com.google.common.truth.Truth.assertThat;
import static com.jesperancinha.test.objects.bufferedreader.NanoStabilityChecker.getStats;

import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * @author JOAO
 */
public class NanoStabilityCheckerTest {
	private static final int NUMBER_OF_TEST_CALLS = 1000000;

	private static final Logger logger = LoggerFactory.getLogger(NanoStabilityChecker.class);

	@Test
	public void testPrintElementTo_100000_elements() {
		final NanoStabilityChecker checker = new NanoStabilityChecker();
		final String testStringFirst = "Test string-" + Math.random();

		final Stats stats = getStats(NUMBER_OF_TEST_CALLS, testStringFirst, checker);

		for (int i = 1; i <= NUMBER_OF_TEST_CALLS; i++) {
			final String testString = "Test string" + Math.random();
			final long resultTimeForThisCheck = checker.printElementTo(testString, System.out);
			logger.info("Time for check {} is {}", i, resultTimeForThisCheck);
			testInToleranceRange(resultTimeForThisCheck, stats);
		}
	}

	private void testInToleranceRange(long resultTimeForThisCheck, Stats stats) {
		long maxValue = stats.getMax();
		long minValue = stats.getMin();
		assertThat(resultTimeForThisCheck).isAtLeast(minValue);
		assertThat(resultTimeForThisCheck).isAtMost(maxValue);
		logger.info(String.format("Test passed for check between %d and %d", minValue, maxValue));
	}

}

File: performance-buffered-reader/src/test/kotlin/com/jesperancinha/test/objects/bufferedreader/NanoStabilityCheckerTest.kt
package com.jesperancinha.test.objects.bufferedreader

import com.google.common.truth.Truth
import com.jesperancinha.test.objects.bufferedreader.NanoStabilityChecker
import com.jesperancinha.test.objects.bufferedreader.NanoStabilityChecker.Companion.getStats
import org.junit.jupiter.api.Test
import org.slf4j.LoggerFactory

/**
 * @author JOAO
 */
class NanoStabilityCheckerTest {
    @Test
    fun testPrintElementTo_100000_elements() {
        val checker = NanoStabilityChecker()
        val testStringFirst = "Test string-" + Math.random()
        val stats = getStats(NUMBER_OF_TEST_CALLS, testStringFirst, checker)
        val testString = "Test string" + Math.random()
        val resultTimeForThisCheck = checker.printElementTo(testString, System.out)
        logger.info("Time for check is {}", resultTimeForThisCheck)
        testInToleranceRange(resultTimeForThisCheck, stats)
    }

    private fun testInToleranceRange(resultTimeForThisCheck: Long, stats: Stats) {
        val maxValue = stats.max
        val minValue = stats.min
        Truth.assertThat(resultTimeForThisCheck).isAtLeast(minValue)
        Truth.assertThat(resultTimeForThisCheck).isAtMost(maxValue)
        logger.info(String.format("Test passed for check between %d and %d", minValue, maxValue))
    }

    companion object {
        private const val NUMBER_OF_TEST_CALLS = 100000
        private val logger = LoggerFactory.getLogger(NanoStabilityChecker::class.java)
    }
}